homeassistant/components/overkiz/__init__.py

```diff
@@ -0,0 +1,123 @@
+"""The Overkiz (by Somfy) integration."""
+import logging
+from typing import NamedTuple
+
+from aiohttp import ClientError, ServerDisconnectedError
+from pyoverkiz.client import OverkizClient
+from pyoverkiz.const import SUPPORTED_SERVERS
+from pyoverkiz.exceptions import (
+    BadCredentialsException,
+    MaintenanceException,
+    TooManyRequestsException,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.aiohttp_client import async_create_clientsession
+
+from .const import (
+    CONF_HUB,
+    DOMAIN,
+    PLATFORMS,
+    UPDATE_INTERVAL,
+    UPDATE_INTERVAL_ALL_ASSUMED_STATE,
+)
+from .coordinator import OverkizDataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class HomeAssistantOverkizData(NamedTuple):
+    """Overkiz data stored in the Home Assistant data object."""
+
+    coordinator: OverkizDataUpdateCoordinator
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Overkiz from a config entry."""
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    server = SUPPORTED_SERVERS[entry.data[CONF_HUB]]
+
+    # To allow users with multiple accounts/hubs, we create a new session so they have separate cookies
+    session = async_create_clientsession(hass)
+    client = OverkizClient(
+        username=username, password=password, session=session, server=server
+    )
+
+    try:
+        await client.login()
+        setup = await client.get_setup()
+    except BadCredentialsException as exception:
+        raise ConfigEntryNotReady("Invalid authentication") from exception
```

Comment:
```suggestion
        _LOGGER.error("Invalid authentication: %s", exception)
        return False
```

This can't get better on retry.

In a future pr `raise ConfigEntryAuthFailed` to start a reauth flow

Reply:
The issue we currently face is that retrying actually works. (context: https://github.com/iMicknl/ha-tahoma/issues/599#issuecomment-995825550)

However I agree that this is not normal behaviour and I will change this to your version, until we reimplement the Config Flow in a follow up PR. Hopefully Somfy will soon fix their issues regarding throwing 'BadCredentials' errors while they have a busy server...

Reply:
Implemented suggestion for now.